llama-cpp-2/src/context/kv_cache.rs

```diff
@@ -31,4 +54,181 @@ impl LlamaContext<'_> {
+    pub fn get_kv_cache_used_cells(&self) -> i32 {
+        unsafe { llama_cpp_sys_2::llama_get_kv_cache_used_cells(self.context.as_ptr()) }
+    }
+
+    /// Clear the KV cache
+    pub fn clear_kv_cache(&mut self) {
+        unsafe { llama_cpp_sys_2::llama_kv_cache_clear(self.context.as_ptr()) }
+    }
+
+    /// Removes all tokens that do not belong to the specified sequence
+    ///
+    /// # Parameters
+    ///
+    /// * `seq_id` - The sequence id to keep
+    pub fn llama_kv_cache_seq_keep(&mut self, seq_id: i32) {
+        unsafe { llama_cpp_sys_2::llama_kv_cache_seq_keep(self.context.as_ptr(), seq_id) }
+    }
+
+    /// Adds relative position "delta" to all tokens that belong to the specified sequence and have positions in [p0, p1)
+    /// If the KV cache is RoPEd, the KV data is updated accordingly:
+    ///   - lazily on next llama_decode()
+    ///   - explicitly with llama_kv_cache_update()
+    ///
+    /// # Parameters
+    ///
+    /// * `seq_id` - The sequence id to update
+    /// * `p0` - The start position of the cache to update. If `None`, the entire cache is updated up to [p1].
+    /// * `p1` - The end position of the cache to update. If `None`, the entire cache is updated starting from [p0].
+    /// * `delta` - The relative position to add to the tokens
+    pub fn kv_cache_seq_add(&mut self, seq_id: i32, p0: Option<u16>, p1: Option<u16>, delta: i32) {
+        unsafe {
+            llama_cpp_sys_2::llama_kv_cache_seq_add(
+                self.context.as_ptr(),
+                seq_id,
+                p0.map_or(-1, i32::from),
+                p1.map_or(-1, i32::from),
+                delta,
+            )
+        }
+    }
+
+    /// Integer division of the positions by factor of `d > 1`
+    /// If the KV cache is RoPEd, the KV data is updated accordingly:
+    ///   - lazily on next llama_decode()
+    ///   - explicitly with llama_kv_cache_update()
+    ///
+    /// # Parameters
+    ///
+    /// * `seq_id` - The sequence id to update
+    /// * `p0` - The start position of the cache to update. If `None`, the entire cache is updated up to [p1].
+    /// * `p1` - The end position of the cache to update. If `None`, the entire cache is updated starting from [p0].
+    /// * `d` - The factor to divide the positions by
+    pub fn kv_cache_seq_div(&mut self, seq_id: i32, p0: Option<u16>, p1: Option<u16>, d: NonZeroU8) {
+        unsafe {
+            llama_cpp_sys_2::llama_kv_cache_seq_div(
+                self.context.as_ptr(),
+                seq_id,
+                p0.map_or(-1, i32::from),
+                p1.map_or(-1, i32::from),
+                d.get() as i32,
+            )
+        }
+    }
+
+    /// Returns the largest position present in the KV cache for the specified sequence
+    ///
+    /// # Parameters
+    ///
+    /// * `seq_id` - The sequence id to get the max position for
+    pub fn kv_cache_seq_pos_max(&self, seq_id: i32) -> i32 {
```

Comment:
What does a negative seq_id mean here (and elsewhere) Is it ever valid / meaningful?

Reply:
As far as I understand, `seq_id` is just a label, ultimately used in a `std::set`, so yes negative numbers are fine.

In the C++ code almost all such parameters actually refers to `typedef int32_t llama_seq_id`, however the existing Rust wrapper does not introduce a newtype (notably `LlamaBatch::add`).

Also, I don't think there is enough checking in the C++ code to check what is passed in makes sense (a proper check would be highly non-trivial and slow). Worse, passing in the wrong thing is not guaranteed to trigger anything noticable. So I'm not sure what is the best way to do in the Rust part.